src/compiler/core.ts

```diff
@@ -2882,3 +2882,56 @@ export function isNodeLikeSystem(): boolean {
         && !(process as any).browser
         && typeof module === "object";
 }
+
+
+/** @internal */
+export interface StackSet<T extends {}> {
+    /**
+     * Returns true if the value has already been pushed.
+     */
+    has(value: T): boolean;
+    /**
+     * Pushes to the stack. The value pushed must not have been pushed before.
+     */
+    push(value: T): void;
+    /**
+     * This pops from the stack until the callback returns true.
+     *
+     * Note that a value is popped, _then_ the callback is evaluated, meaning
+     * that if you return true to stop iteration, the value that the callback
+     * returned true for will still have been popped.
+     */
+    popAll(callback: (v: T) => boolean): void;
+}
+
+/** @internal */
+export function createStackSet<T extends {}>(): StackSet<T> {
+    // Why var? It avoids TDZ checks in the runtime which can be costly.
+    // See: https://github.com/microsoft/TypeScript/issues/52924
+    /* eslint-disable no-var */
+    var set = new Set<T>();
+    var stack: T[] = [];
+    var end = 0;
+    /* eslint-enable no-var */
+
+    return {
+        has(value) {
+            return set.has(value);

```

Comment:
A part of me thinks that we could use just the array with linear searching up until a limit, rather than always using the set, but this PR appears to be an improvement regardless of that speculative optimization.

Reply:
Tested and it doesn't seem like using a list is actually faster even when checking one thing. That or my microbenchmark is bad.